Bio/PDB/rotamers.py

```diff
@@ -0,0 +1,279 @@
+"""
+A tool that introduces a point mutation to a given residue based on the Dunbrack rotamer library.
+The tool uses sample residues stored in small PDB files, and uses a reduced rotamer library
+to save space (only rotamers with higher than 5% probability are present).
+The tool makes in place modification on a PDB structure object.
+"""
+
+from Bio.Entrez.Parser import ValidationError
+from Bio.PDB import Polypeptide
+from Bio.PDB.NeighborSearch import NeighborSearch
+from Bio.PDB.Selection import unfold_entities
+from Bio.PDB.vectors import calc_dihedral, Vector, rotaxis2m
+from Bio import SVDSuperimposer
+from collections import defaultdict
+from Bio.PDB.ic_data import CHI_ANGLES, RESIDUE_ORDER, VW_RADII
+import numpy as np
+import os
+import gzip
+
+
+DATA_DIR = os.path.join(os.path.dirname(os.path.realpath(__file__)), "rotamers")
+DUNBRACK_FILTERED = f"{DATA_DIR}/dunbrack_reduced.gz"
+
+
+class Rotamer:
+    """
+    Object to store rotamers
+    """
+
+    def __init__(self, resname, psi, phi, chi1, chi2, chi3, chi4, probability):
+        """
+        Init the rotamer with given diherdral angles
+        """
+        self.psi = float(psi)
+        self.phi = float(phi)
+        self.chi1 = float(chi1)
+        self.chi2 = float(chi2)
+        self.chi3 = float(chi3)
+        self.chi4 = float(chi4)
+        self.resname = resname
+        self.probability = float(probability)
+        self.sample_residue = False
+
+    def __getitem__(self, item):
+        """
+        Returns the given angle property of the rotamer
+        """
+        return getattr(self, item.lower())
+
+    def set_sample_residue(self, sample_residue):
+        """
+        Sets a sample residue for the rotamer object
+        This function is only called during RotamerSampling to save memory
+        """
+        self.sample_residue = sample_residue
+
+    def get_sample_residue(self):
+        """
+        Returns the stored sample residue
+        """
+        if self.sample_residue:
+            return self.sample_residue
+        raise ValueError("Sample residue is not set, use RotamerSampling to set it!")
```

Comment:
Explicit getters and setters like this are considered non-pythonic. Some of the older code in the library uses them, but I feel it's probably best to avoid them in new code. If you wanted to do this in a pythonic way, you could create sample_residue as a property like:
```
@property
def sample_residue(self, sample_residue):
    if self._sample_residue:
        return self._sample_residue
    raise ValueError("Sample residue is not set, use RotamerSampling to set it!")

@sample_residue.setter
def sample_residue(self, sample_residue):
    self._sample_residue = sample_residue
```

Similarly I think overloading `__getitem__` like this is probably not best unless there is a particular good reason. The philosophy of Python is to have only one best way of doing something when possible. Giving this class two ways of accessing the same attribute violates this, and makes it slightly harder to understand what is going on in the code

But honestly I'm not sure if it's a good idea it is to increase the complexity have having nested classes though. You could probably just set this up as a named tuple, have None represent unavailable sample_residue, and handle conversions and whatnot within RotamerSampling to avoid the user having to search around for what each class does
```
class Rotamer(typing.NamedTuple):
    """
    Object to store rotamers
    """
    psi: float
    phi: float
    chi1: float
    chi2: float
    chi3: float
    chi4: float
    resname: str # ?
    probability: float
    sample_residue: typing.Optional[dict] = None # ?
```

Reply:
Thinking about it some more, you’re probably right to keep it as a class. The caching with sample_residue would be somewhat less natural if it were immutable like a NamedTuple. There are other options like TypedDict, but a standard class is probably more natural